Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm

```diff
@@ -569,17 +515,18 @@
     processSampleBuffers();
 }
 
-void AudioSampleBufferCompressor::addSampleBuffer(CMSampleBufferRef buffer)
+Ref<GenericPromise> AudioSampleBufferCompressor::addSampleBuffer(CMSampleBufferRef buffer)
 {
     // Heap allocations are forbidden on the audio thread for performance reasons so we need to
     // explicitly allow the following allocation(s).
     DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
 
-    m_serialDispatchQueue->dispatch([weakThis = ThreadSafeWeakPtr { *this }, buffer = RetainPtr { buffer }] {
+    return invokeAsync(m_serialDispatchQueue, [weakThis = ThreadSafeWeakPtr { *this }, buffer = RetainPtr { buffer }] {
         RefPtr protectedThis = weakThis.get();
         if (!protectedThis || !protectedThis->m_isEncoding)
-            return;
+            return GenericPromise::createAndReject();
         protectedThis->processSampleBuffer(buffer.get());
+        return std::exchange(protectedThis->m_lastError, noErr) ? GenericPromise::createAndReject() : GenericPromise::createAndResolve();
```

Comment:
When there is an error, this is usually terminal for WebCodecs, and there is a need to reconfigure the encoder, which would then create a new internal encoder.
Do we need to std::exchange?

Reply:
it prevents having to deal with what to do when the converter is in an error state.